draft-ietf-dnsop-dns-catalog-zones.md

```diff
@@ -384,22 +386,30 @@ Note that secondary nameservers may attempt to transfer the catalog zone upon
 refresh timeout, so care must be taken to make the member zones available
 before any update to the list of member zones is visible in the catalog zone.
 
-When zones are deleted from a catalog zone, a primary MAY delete the member
-zone immediately after notifying secondaries.  It is up to the secondary
-nameserver to handle this condition correctly.
+## Resetting a zone's associated state {#zonereset}
 
-{#zonereset}
 When the `<m-unique-N>` label of a member zone changes, all its associated state MUST be reset, including the zone itself.
 This can be relevant for example when zone ownership is changed.
 In that case one does not want the new owner to inherit the metadata.
 Other situations might be resetting DNSSEC state, or forcing a new zone transfer.
 A simple removal followed by an addition of the member zone would be insufficient for this purpose because it is infeasible for secondaries to track, due to missed notifies or being offline during the removal/addition.
 
-# Updating Catalog Zones
+## Deleting and "moving" member zones
+
+When zones are deleted from a catalog zone, a primary MAY delete the member
+zone immediately after notifying secondaries.
+
+When a zone is deleted from a catalog zone, a secondary MUST NOT remove the zone and associated state data if the zone was not configured from that specific catalog zone in the first place.
+Only when the zone was configured from a specific catalog zone, and the zone is removed as a member from that specific catalog zone, the zone and associated state MUST be removed from the secondary.
+
+Subsequently, after removal, a secondary MUST check if another catalog zone is present which has the zone as a member.
+In that case the zone must be configured and associated according to the set of configuration associated with that catalog zone.
+If there are multiple other catalog zones which have the previously removed zone as member, one should be picked randomly.
 
-TBD: Explain updating catalog zones using DNS UPDATE.
+If a member zone, after removal, is associated with another catalog zone, and the `<m-unique-N>` label of the member zone remains the same, the associated state data MUST NOT be removed.
+This is to facilitate "moving" a member zone from one catalog zone to another without having to retransfer the zone content or reestablish other associated state.
```

Comment:
I think the ordering on these steps (403-406) might be backward.

As written, it says that FIRST the zone must be removed from the secondary, and SECOND (after removal) the secondary must check if the zone is associated with other catalog zones. Is that right? To me, those steps seem logically backward, but perhaps I read them incorrectly.

I would think that a secondary, having been notified that the primary removed a zone from a specific catalog zone, should check for other catalog zone associations before removing the zone from itself.

For 409-410, the wording suggests that the zone is removed prior to checks being run, and if those checks pass, it "must not" be removed. So, it must not be removed after being removed. This needs further clarification.

Reply:
In addition to that, we should write some words suggesting (I think MUST is too strong, because not everybody may be able to implement that easily) that transitions should happen in an atomic manner where possible.